libraries/rush-lib/src/cli/actions/PublishAction.ts

```diff
@@ -337,50 +337,53 @@ export class PublishAction extends BaseRushAction {
         this._setDependenciesBeforeCommit();
 
         // Create and push appropriate Git tags.
-        this._gitAddTags(publishGit, orderedChanges);
-        publishGit.push(tempBranchName, !this._ignoreGitHooksParameter.value);
+        await this._gitAddTagsAsync(publishGit, orderedChanges);
+        await publishGit.pushAsync(tempBranchName, !this._ignoreGitHooksParameter.value);
 
         // Now merge to target branch.
-        publishGit.checkout(this._targetBranch.value!);
-        publishGit.pull(!this._ignoreGitHooksParameter.value);
-        publishGit.merge(tempBranchName, !this._ignoreGitHooksParameter.value);
-        publishGit.push(this._targetBranch.value!, !this._ignoreGitHooksParameter.value);
-        publishGit.deleteBranch(tempBranchName, true, !this._ignoreGitHooksParameter.value);
+        await publishGit.checkoutAsync(this._targetBranch.value!);
+        await publishGit.pullAsync(!this._ignoreGitHooksParameter.value);
+        await publishGit.mergeAsync(tempBranchName, !this._ignoreGitHooksParameter.value);
+        await publishGit.pushAsync(this._targetBranch.value!, !this._ignoreGitHooksParameter.value);
+        await publishGit.deleteBranchAsync(tempBranchName, true, !this._ignoreGitHooksParameter.value);
       } else {
-        publishGit.checkout(this._targetBranch.value!);
-        publishGit.deleteBranch(tempBranchName, false, !this._ignoreGitHooksParameter.value);
+        await publishGit.checkoutAsync(this._targetBranch.value!);
+        await publishGit.deleteBranchAsync(tempBranchName, false, !this._ignoreGitHooksParameter.value);
       }
     }
   }
 
-  private _publishAll(git: PublishGit, allPackages: ReadonlyMap<string, RushConfigurationProject>): void {
+  private async _publishAllAsync(
+    git: PublishGit,
+    allPackages: Map<string, RushConfigurationProject>
+  ): Promise<void> {
     // eslint-disable-next-line no-console
     console.log(`Rush publish starts with includeAll and version policy ${this._versionPolicy.value}`);
 
     let updated: boolean = false;
 
-    allPackages.forEach((packageConfig, packageName) => {
+    for (const [packageName, packageConfig] of allPackages) {
```

Comment:
Async.forEachAsync? Though not sure what issues multiple async git commands may cause

Reply:
Yeah that feels a little unsafe. Let's leave this sequential.